In [1]:
install.packages('ggplot2')
#install.packages('caret')

also installing the dependencies 'colorspace', 'cli', 'utf8', 'nlme', 'farver', 'labeling', 'lifecycle', 'munsell', 'R6', 'RColorBrewer', 'viridisLite', 'fansi', 'magrittr', 'pillar', 'pkgconfig', 'vctrs', 'glue', 'gtable', 'isoband', 'mgcv', 'rlang', 'scales', 'tibble', 'withr'




  There are binary versions available but the source versions are later:
              binary  source needs_compilation
colorspace     2.0-1   2.0-3              TRUE
cli            2.5.0   3.4.1              TRUE
utf8           1.2.1   1.2.2              TRUE
nlme         3.1-152 3.1-159              TRUE
farver         2.1.0   2.1.1              TRUE
lifecycle      1.0.0   1.0.2             FALSE
R6             2.5.0   2.5.1             FALSE
RColorBrewer   1.1-2   1.1-3             FALSE
viridisLite    0.4.0   0.4.1             FALSE
fansi          0.4.2   1.0.3              TRUE
magrittr       2.0.1   2.0.3              TRUE
pillar         1.6.0   1.8.1             FALSE
vctrs          0.3.8   0.4.1              TRUE
glue           1.4.2   1.6.2              TRUE
gtable         0.3.0   0.3.1             FALSE
isoband        0.2.4   0.2.5              TRUE
mgcv          1.8-35  1.8-40              TRUE
rlang         0.4.11   1.0.6              TRUE
scales         1.1.1   1.2.1     

installing the source packages 'lifecycle', 'R6', 'RColorBrewer', 'viridisLite', 'pillar', 'gtable', 'scales', 'withr', 'ggplot2'

Warning message in install.packages("ggplot2"):
"installation of package 'lifecycle' had non-zero exit status"Warning message in install.packages("ggplot2"):
"installation of package 'pillar' had non-zero exit status"Warning message in install.packages("ggplot2"):
"installation of package 'scales' had non-zero exit status"Warning message in install.packages("ggplot2"):
"installation of package 'ggplot2' had non-zero exit status"

In [5]:
#install.packages("class")


  There is a binary version available but the source version is later:
      binary source needs_compilation
class 7.3-19 7.3-20              TRUE

  Binaries will be installed
package 'class' successfully unpacked and MD5 sums checked

The downloaded binary packages are in
	C:\Users\amara\AppData\Local\Temp\RtmpUbAAOi\downloaded_packages


In [6]:
library("MASS")
#library("R.utils")
#library('caret')
library("class")
#library("ggplot2")
library(ISLR)
library(MASS)
library(glmnet)
library(caTools)

Warning message:
"package 'class' was built under R version 3.6.3"Warning message:
"package 'glmnet' was built under R version 3.6.3"Loading required package: Matrix
Warning message:
"package 'Matrix' was built under R version 3.6.3"Loaded glmnet 4.1-1
Warning message:
"package 'caTools' was built under R version 3.6.3"

## 3 ) ESL textbook exercise 2.8 modified: Compare the classification
## performance of linear regression and k-nearest neighbor classification on the
## zipcode data. In particular, consider only the 7’s and 9’s for this problem, and k =
## 1, 3, 5, 7, 9, 11, 13,15. Show the test error for each choice of k. Describe your
## results – are you surprised by the differences in performance? 

### lets load the two data sets

In [7]:
load("C:/Users/amara/OneDrive/Desktop/SDM1/assignment2/zip.test.RData")
head(zip.test)
test_dats = zip.test
load("C:/Users/amara/OneDrive/Desktop/SDM1/assignment2/zip.train.RData")
head(zip.train)
train_dats = zip.train

[,1] [,2] [,3] [,4] [,5]   [,6] [,7]   [,8]   [,9]   [,10]  [,11] [,12] 
[1,] 9    -1   -1   -1   -1.000 -1.0 -0.948 -0.561  0.148  0.384 ...   -1.000
[2,] 6    -1   -1   -1   -1.000 -1.0 -1.000 -1.000 -1.000 -1.000 ...   -1.000
[3,] 3    -1   -1   -1   -0.593  0.7  1.000  1.000  1.000  1.000 ...    1.000
[4,] 6    -1   -1   -1   -1.000 -1.0 -1.000 -1.000 -1.000 -1.000 ...   -1.000
[5,] 6    -1   -1   -1   -1.000 -1.0 -1.000 -1.000 -0.858 -0.106 ...    0.901
[6,] 0    -1   -1   -1   -1.000 -1.0 -1.000  0.195  1.000  0.054 ...    0.224
     [,13]  [,14]  [,15]  [,16]  [,17]  [,18]  [,19] [,20] [,21]
[1,] -0.908  0.430  0.622 -0.973 -1.000 -1.000 -1    -1    -1   
[2,] -1.000 -1.000 -1.000 -1.000 -1.000 -1.000 -1    -1    -1   
[3,]  0.717  0.333  0.162 -0.393 -1.000 -1.000 -1    -1    -1   
[4,] -1.000 -1.000 -1.000 -1.000 -1.000 -1.000 -1    -1    -1   
[5,]  0.901  0.901  0.290 -0.369 -0.867 -1.000 -1    -1    -1   
[6,]  1.000  0.988  0.187  0.139 -0.641 -0.812 -1    -1    -1

[,1] [,2] [,3] [,4] [,5]   [,6]   [,7]   [,8]   [,9]   [,10]  [,11] [,12] 
[1,] 6    -1   -1   -1   -1.000 -1.000 -1.000 -1.000 -0.631  0.862 ...    0.304
[2,] 5    -1   -1   -1   -0.813 -0.671 -0.809 -0.887 -0.671 -0.853 ...   -0.671
[3,] 4    -1   -1   -1   -1.000 -1.000 -1.000 -1.000 -1.000 -1.000 ...   -1.000
[4,] 7    -1   -1   -1   -1.000 -1.000 -0.273  0.684  0.960  0.450 ...   -0.318
[5,] 3    -1   -1   -1   -1.000 -1.000 -0.928 -0.204  0.751  0.466 ...    0.466
[6,] 6    -1   -1   -1   -1.000 -1.000 -0.397  0.983 -0.535 -1.000 ...    0.803
     [,13]  [,14]  [,15]  [,16]  [,17]  [,18]  [,19]  [,20]  [,21]
[1,]  0.823  1.000  0.482 -0.474 -0.991 -1.000 -1.000 -1.000 -1   
[2,] -0.671 -0.033  0.761  0.762  0.126 -0.095 -0.671 -0.828 -1   
[3,] -1.000 -1.000 -0.109  1.000 -0.179 -1.000 -1.000 -1.000 -1   
[4,]  1.000  0.536 -0.987 -1.000 -1.000 -1.000 -1.000 -1.000 -1   
[5,]  0.639  1.000  1.000  0.791  0.439 -0.199 -0.883 -1.000 -1   
[6,]  0.015 -0.862 -0.871 -0.437 -1.000 -1.000 -1.000 -1.000 -1

## We only need to consider the 7's and the 9's

In [8]:
train_data = train_dats[which(zip.train[,1] == 7 | zip.train[,1] == 9),]
test_data = test_dats[which(zip.test[,1] ==7 | zip.test[,1] == 9),]

Obtained Data that only has 7's and the 9's

## Lets convert our data into a data frame 

## We are making the data points replacing 7,9 with 0 and 1 to make it easier to classify

In [9]:
train_data[which(train_data[,1] == 7),1] = 0
train_data[which(train_data[,1] == 9),1] = 1

test_data[which(test_data[,1] == 7),1] = 0
test_data[which(test_data[,1] == 9),1] = 1

Create Data Frame

In [10]:
data_train=data.frame(train_data)
data_train
data_test=data.frame(test_data)
data_test

X1  X2     X3     X4     X5     X6     X7     X8     X9     X10    ...
1    0   -1.000 -1.000 -1.000 -1.000 -1.000 -0.273  0.684  0.960  0.450 ...
2    0   -1.000 -1.000 -1.000 -1.000 -1.000 -0.596  0.912  1.000  0.290 ...
3    0   -1.000 -1.000 -1.000 -1.000 -1.000 -1.000 -1.000 -1.000 -0.632 ...
4    0   -1.000 -1.000 -1.000 -0.929  0.351  0.798  0.806  0.114  0.015 ...
5    0   -1.000 -1.000 -1.000 -1.000 -0.869  0.777 -0.007 -0.697 -1.000 ...
6    0   -1.000 -1.000 -1.000 -1.000 -0.846 -0.011 -0.920 -1.000 -1.000 ...
7    0   -1.000 -1.000 -1.000 -1.000 -1.000 -1.000 -1.000 -1.000 -0.959 ...
8    0   -1.000 -1.000 -1.000 -0.027  0.395 -0.948 -1.000 -1.000 -1.000 ...
9    0   -0.201  0.752  1.000  1.000  0.793  0.515  0.515  0.765  0.970 ...
10   0   -1.000 -1.000 -1.000 -1.000  0.781 -0.728 -1.000 -1.000 -0.918 ...
11   0   -1.000 -1.000 -1.000 -1.000 -1.000 -1.000 -0.927 -0.449 -0.168 ...
12   0   -1.000 -1.000 -1.000 -1.000 -1.000 -0.820 -0.190  0.457  0.881 ...
13   0   -1.000 -1.000 -1.000 -1.000 -1.000 -0.726 -0.325 -0.263 -0.263 ...
14   0   -1.000 -1.000 -1.000 -1.000 -1.000 -1.000 -0.665 -0.429 -0.032 ...
15   1   -1.000 -1.000 -1.000 -1.000 -1.000 -1.000 -0.988  0.450  0.226 ...
16   1   -1.000 -1.000 -1.000 -1.000 -0.999 -0.491  0.384  0.826  0.498 ...
17   1   -1.000 -1.000 -1.000 -1.000 -1.000 -1.000 -1.000 -0.912 -0.519 ...
18   0   -1.000 -1.000 -1.000 -0.475 -0.392  0.232  0.964  1.000  1.000 ...
19   1   -1.000 -1.000 -1.000 -1.000 -1.000 -1.000 -1.000 -0.905 -0.261 ...
20   0   -1.000 -1.000 -0.658  0.987  0.323 -0.295 -0.362 -0.009  0.319 ...
21   1   -1.000 -1.000 -1.000 -1.000 -1.000 -1.000 -0.719  0.123  0.617 ...
22   1   -1.000 -1.000 -1.000 -0.999 -0.713 -0.060  0.059  0.178  0.721 ...
23   1   -1.000 -1.000 -1.000 -0.938 -0.489  0.125  0.675  1.000  1.000 ...
24   1   -1.000 -1.000 -1.000 -1.000 -1.000 -1.000 -0.991  0.019  0.955 ...
25   0   -1.000 -0.101  1.000  0.921  0.307 -0.477 -0.746 -0.746 -0.966 ...
26   1   -1.000 -1.000 -1.000 -1.000 -1.000 -1.000 -0.529  0.275  0.801 ...
27   0   -1.000 -1.000 -1.000 -0.945 -0.943 -0.943 -0.965 -0.966 -0.943 ...
28   1   -1.000 -1.000 -1.000 -1.000 -1.000 -1.000 -1.000 -0.751  0.065 ...
29   1   -1.000 -1.000 -1.000 -1.000 -0.993 -0.547  0.721  1.000  0.791 ...
30   1   -1.000 -1.000 -1.000 -1.000 -1.000 -1.000 -0.409  0.565  1.000 ...
...  ... ...    ...    ...    ...    ...    ...    ...    ...    ...       
1260 0   -1.000 -1.000 -1.000 -1.000 -1.000 -0.608  0.654  0.361 -1.000 ...
1261 1   -1.000 -1.000 -1.000 -1.000 -1.000 -1.000 -0.912 -0.261  0.414 ...
1262 1   -1.000 -1.000 -1.000 -1.000 -0.901  0.397  1.000  1.000  1.000 ...
1263 1   -1.000 -1.000 -1.000 -1.000 -0.998 -0.368 -0.046  0.459  0.698 ...
1264 1   -1.000 -1.000 -1.000 -1.000 -0.929 -0.435  0.259  0.638  0.869 ...
1265 1   -1.000 -1.000 -1.000 -1.000 -1.000 -0.965 -0.496  0.140  0.958 ...
1266 0   -1.000 -1.000 -1.000 -1.000 -1.000 -1.000 -1.000 -1.000 -1.000 ...
1267 0   -1.000 -1.000 -1.000 -0.531  0.442  0.457  0.457  0.746  1.000 ...
1268 0   -1.000 -1.000 -1.000 -1.000 -0.907 -0.574 -0.494 -0.049  0.409 ...
1269 0   -1.000 -0.929  0.000  0.403  0.529  0.944  0.998  0.669  0.612 ...
1270 0   -1.000 -1.000 -1.000 -1.000 -0.961 -0.431  0.368  0.793  1.000 ...
1271 1   -1.000 -1.000 -1.000 -1.000 -0.962  0.089  0.684  1.000  1.000 ...
1272 0   -1.000 -1.000 -1.000 -1.000 -1.000 -1.000 -1.000 -1.000 -0.961 ...
1273 1   -1.000 -1.000 -1.000 -1.000 -1.000 -1.000 -0.940  0.155  0.845 ...
1274 0   -1.000 -1.000 -0.969  0.204  0.573  0.573  0.573  0.594  1.000 ...
1275 0   -1.000 -1.000 -1.000 -1.000 -0.718  0.157  0.289  0.289  0.504 ...
1276 0   -0.981 -0.418  0.002  0.275 -0.830 -0.883 -0.883 -0.883 -0.750 ...
1277 0   -1.000 -1.000 -0.417  0.881  0.215 -0.535 -0.746 -0.746 -0.746 ...
1278 0   -1.000 -1.000 -1.000 -1.000 -1.000 -0.765 -0.477 -0.477 -0.137 ...
1279 0   -1.000 -1.000 -1.000 -1.000 -1.000 -1.000 -0.878 -1.000 -1.000 ...
1280 1   -1.000 -1.000 -1.000 -1.000 -1.000 -0.494  0

X1  X2     X3     X4     X5     X6     X7     X8     X9     X10    ...
1   1   -1     -1.000 -1.000 -1.000 -1.000 -0.948 -0.561  0.148  0.384 ...
2   1   -1     -1.000 -1.000 -1.000 -1.000 -0.904 -0.060  0.638  0.678 ...
3   1   -1     -1.000 -1.000 -0.916 -0.600 -0.034  0.711  1.000  0.488 ...
4   1   -1     -1.000 -1.000 -1.000 -1.000 -1.000 -1.000 -0.441  0.530 ...
5   0   -1     -0.990  0.700  0.283 -0.131 -0.598 -1.000 -1.000 -1.000 ...
6   0   -1     -1.000 -0.862  0.000  0.296 -0.231  0.142  0.384  0.763 ...
7   1   -1     -1.000 -1.000 -1.000 -1.000 -1.000 -0.973 -0.516  0.406 ...
8   0   -1     -1.000 -1.000 -1.000 -1.000 -0.930 -0.524 -0.524 -0.129 ...
9   0   -1     -0.798  0.219  0.219  0.799  0.249  0.219  0.219  0.219 ...
10  1   -1     -1.000 -1.000 -1.000 -1.000 -0.743  0.272  0.948  1.000 ...
11  1   -1     -1.000 -1.000 -1.000 -1.000 -1.000 -0.412  0.272  0.272 ...
12  1   -1     -1.000 -1.000 -1.000 -1.000 -0.891 -0.025  0.757  1.000 ...
13  0   -1     -1.000 -0.634  0.162 -0.027  0.612  1.000  0.741 -0.416 ...
14  1   -1     -1.000 -1.000 -1.000 -1.000 -1.000 -0.870 -0.252  0.630 ...
15  0   -1     -1.000 -1.000 -1.000 -0.986  0.068  0.200  0.279 -0.366 ...
16  0   -1     -0.296  1.000  0.950  0.148 -0.502 -0.776 -1.000 -1.000 ...
17  1   -1     -1.000 -1.000 -1.000 -1.000 -0.745  0.137  0.996  0.976 ...
18  0   -1     -1.000  0.190  1.000  0.871  0.402  0.998  1.000  1.000 ...
19  1   -1     -1.000 -1.000 -1.000 -1.000 -1.000 -0.568 -0.036  0.787 ...
20  0   -1     -1.000 -0.505  0.919 -0.037 -0.467 -0.778 -0.778 -0.778 ...
21  0   -1     -1.000 -1.000 -1.000 -0.360  0.999  0.176 -0.603 -0.826 ...
22  1   -1     -1.000 -1.000 -1.000 -1.000 -0.641 -0.027  0.736  0.893 ...
23  1   -1     -1.000 -1.000 -1.000 -1.000 -0.977  0.003  0.944  1.000 ...
24  1   -1     -1.000 -1.000 -1.000 -1.000 -0.997 -0.449  0.574  0.953 ...
25  0   -1     -1.000 -1.000 -1.000 -1.000 -0.820 -0.083  0.746  1.000 ...
26  1   -1     -1.000 -1.000 -1.000 -1.000 -1.000 -0.892  0.149  0.868 ...
27  1   -1     -1.000 -1.000 -1.000 -1.000 -1.000 -0.590  0.350  0.991 ...
28  1   -1     -1.000 -0.925  0.013  0.457  0.653  0.556  0.234 -0.718 ...
29  0   -1     -1.000 -1.000 -1.000 -1.000 -1.000 -1.000 -1.000 -0.969 ...
30  0   -1     -1.000 -0.966 -0.144 -0.487 -0.768 -1.000 -0.992 -0.600 ...
... ... ...    ...    ...    ...    ...    ...    ...    ...    ...       
295 1   -1.000 -1.000 -1.000 -1.000 -1.000 -1.000 -1.000 -0.906  0.268 ...
296 1   -1.000 -1.000 -1.000 -1.000 -0.965 -0.450  0.185  0.650  1.000 ...
297 1   -1.000 -1.000 -1.000 -1.000 -1.000 -1.000 -1.000 -0.669  0.697 ...
298 1   -0.858  0.731  0.542  0.525  1.000 -0.598  0.284  0.800 -0.871 ...
299 0   -1.000 -1.000 -1.000  0.212  0.875  1.000  1.000  1.000  1.000 ...
300 0   -1.000 -1.000 -1.000 -1.000 -1.000 -0.738 -0.055  0.200  0.953 ...
301 1   -1.000 -1.000 -1.000 -1.000 -1.000 -1.000 -1.000 -0.747  0.293 ...
302 0   -1.000 -1.000 -0.788  0.133  0.135  0.786  1.000  1.000  1.000 ...
303 0   -1.000 -1.000 -1.000 -1.000 -0.985 -0.392  0.743  0.673 -0.602 ...
304 1   -1.000 -1.000 -1.000 -1.000 -1.000 -0.493  0.226  0.926  1.000 ...
305 1   -1.000 -1.000 -1.000 -1.000 -1.000 -1.000 -0.984 -0.052  0.577 ...
306 1   -1.000 -1.000 -1.000 -1.000 -1.000 -1.000 -1.000 -0.543  0.337 ...
307 1   -1.000 -1.000 -1.000 -1.000 -1.000 -1.000 -0.968 -0.144  0.792 ...
308 0   -1.000 -1.000 -1.000 -0.407  0.893  0.466  0.466  0.466  0.176 ...
309 1   -1.000 -1.000 -1.000 -1.000 -1.000 -0.573  0.199  0.920  1.000 ...
310 1   -1.000 -1.000 -1.000 -1.000 -1.000 -1.000 -0.988 -0.383  0.417 ...
311 0   -1.000 -1.000 -0.723  0.287 -0.420 -0.423 -0.423 -0.666 -0.795 ...
312 0   -1.000 -0.939  0.431  0.978  0.077 -0.770 -0.859 -0.013  0.304 ...
313 1   -1.000 -1.000 -1.000 -1.000 -1.000 -0.583  0.567  0.884  0.357 ...
314 0   -1.000 -1.000 -1.000 -1.000 -1.000 -0.501 -0.085  0.443  0.500 ...
315 1   -1.000 -1.000 -1.000 -1.000 -1.000 -0.719  0.138  0.790  0.545 ...
316 0   -1.000 -0.362  0.200 

## lets apply a linear model

Comparing Column X1 with all the columns in the dataset 

In [11]:
linear_model = lm(X1~.,data = data_train)
y_hat = predict(linear_model ,data_test)
error = mean((data_test[,"X1"] - y_hat)^2)
error

Warning message in predict.lm(linear_model, data_test):
"prediction from a rank-deficient fit may be misleading"

[1] 0.647209

## MSE 64 %
Which is Very high 

# K nearest neighbour

Picking Ks for classification problem 

In [13]:
K_list = c(1,3,5,7,9,11,13,15)

In [14]:
k_error=0

## Predicting for test

In [15]:
require(class)

## We are using as.factor to convert the data into 0s and 1s

In [16]:
fac_test = as.factor(data_test[,1])
fac_train = as.factor(data_train[,1])

## Using for loop to do all the different values of K

In [21]:
library(class)

for (i in 1:8) {
  yhat = knn(data_train, data_test, fac_train, K_list[i])
  k_error[i] = mean(yhat != fac_test)
    #print(i)
}




In [22]:
k_error

[1] 0.02469136 0.02469136 0.02469136 0.02777778 0.03086420 0.03395062 0.03703704
[8] 0.03395062

## We See that the linear regression model has a very high error which is true due to the fact that linear regression is not good at classification problems and this being a classification problem gives a high error with linear regression where as kNN performs really well with errors rangeing between 2.4%-3.7% which is good and shows that K 